dedalus/core/solvers.py

```diff
@@ -133,22 +152,94 @@ def solve_sparse(self, pencil, N, target, rebuild_coeffs=False, **kw):
         # Solve as sparse general eigenvalue problem
         A = pencil.L_exp
         B = -pencil.M_exp
+        # Solve for the right eigenvectors
         self.eigenvalues, self.eigenvectors = scipy_sparse_eigs(A=A, B=B, N=N, target=target, matsolver=self.matsolver, **kw)
         if pencil.pre_right is not None:
+            # I remember Keaton explaining long ago that this step is only necessary for right eigenvectors, I think
             self.eigenvectors = pencil.pre_right @ self.eigenvectors
+        if left == True:
+            # Solve for the left eigenvectors
+            # Note: this definition of "left eigenvectors" is consistent with the documentation for scipy.linalg.eig
+            self.left_eigenvalues, self.left_eigenvectors = scipy_sparse_eigs(A=A.getH(),
+                                                                              B=B.getH(),
+                                                                              N=N, target=np.conjugate(target),
+                                                                              matsolver=self.matsolver, **kw)
+            # TODO: make it so that left_eigenvectors is Y^* (do the same for eigenvalues)?
+            # TODO: and modified_left_eigenvectors is Y^*M?
+            # TODO: Maybe start with making it consistent with scipy.linalg.eig formalism, with clear docstring, then ask Jeff Oishi for input
+            # TODO: after normalizing, one last iteration to clean up orthogonality? -- kick this can down the road, Keaton will look into it
+            # TODO: decide whether and how to sort eigenvalues and eigenvectors before this point.
+            if not np.allclose(self.eigenvalues, np.conjugate(self.left_eigenvalues)):
```

Comment:
Hey Keaton, Zach Williams (a collaborator and the current number one user of this functionality I suspect) is telling me that, even with sorting, these mismatches are a near constant problem for him. If he passes the `target` kwarg, it's not as bad for the modes near the target, but he's often searching for 10-100 modes at once, and it sounds like the mismatch is a significant obstacle for him.

Any idea what can be done? Is this an inherent consequence of using an iterative solver? Would tinkering with any of the matrix solvers in dedalus.cfg affect this?

Reply:
Is the mismatch occuring because the order is wrong or because the left eigenvalues/vectors are inaccurate? If it's the latter, is the problem just that he doesn't have resolution on those modes?

Reply:
Hmm good question. He's looking into it now. Part of it is certainly the order, but it sounds like he has no problem sorting the modes himself. He finds that, even after sorting, there are eigenvalues in one set that don't correspond to eigenvalues in the other set.

Reply:
Seems like a good argument for me figuring out how to put this into eigentools!

Reply:
Ok, Zach has looked into this. There was an issue where a mistake in the sorting caused apparent mismatches that were eventually cleared up. Even after fixing that, there were still some mismatches, but now we know why. He's looking at EVPs corresponding to the tearing instability where the eigenvalue spectrum is symmetric along one axis in the complex plane (see figure 1 [here](https://arxiv.org/abs/1808.10208) and note the symmetry with respect to reflections over the horizontal axis). So it seems that if your target/guess eigenvalue passed to the solver is equidistant to two different eigenvalues due to this symmetry, the solver sometimes finds one frequency for the right modes and the other frequency for the left modes. At least, that's our working hypothesis for what's going on.